File: 동적계획법/leetcode983(Medium)-Minimum Cost For Tickets.py
# 41ms(87%), 14MB(58%)
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        max_day = max(days)
        dp = [0] * (max_day + 1)
        dayset = set(days)

        def f(day):
            if 1 <= day <= max_day:
                return dp[day]
            return 0

        for day in range(1, max_day + 1):
            if day in dayset:
                dp[day] = min( f(day - 1) + costs[0], f(day - 7) + costs[1], f(day - 30) + costs[2] )
            else:
                dp[day] = dp[day - 1]

        return dp[-1]
    

# 범위를 신경안쓰기 위해, **뒤에서부터** DP를 진행해가는 방법도 가능
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        dp = [0] * (365 + 30 + 1)

        for day in range(365, -1, -1):
            if not day in days:
                dp[day] = dp[day + 1]
                continue
            dp[day] = min(dp[day + 1] + costs[0], dp[day + 7] + costs[1], dp[day + 30] + costs[2])

        return dp[1]

File: 동적계획법/leetcode983(Medium)-Minimum Cost For Tickets.js
// 61ms(80.85%), 43MB(34%)
/**
 * @param {number[]} days
 * @param {number[]} costs
 * @return {number}
 */
var mincostTickets = function (days, costs) {
  const maxDay = Math.max(...days)
  const daySet = new Set(days)
  const dp = new Array(maxDay + 1).fill(0)

  const f = (day) => {
    return 0 <= day && day <= maxDay ? dp[day] : 0
  }

  const [a, b, c] = costs
  for (let day = 1; day < maxDay + 1; day += 1) {
    if (daySet.has(day)) {
      dp[day] = Math.min(f(day - 1) + a, f(day - 7) + b, f(day - 30) + c)
    } else {
      dp[day] = dp[day - 1]
    }
  }

  return dp[maxDay] // == dp[dp.length - 1]
}

// 범위 신경 안쓰려면 뒤에서부터 풀 수도 있음.
/**
 * @param {number[]} days
 * @param {number[]} costs
 * @return {number}
 */
var mincostTickets = function (days, costs) {
  const daySet = new Set(days)
  const dp = new Array(365 + 30 + 1).fill(0)

  const [a, b, c] = costs
  for (let day = 365; day >= 0; day -= 1) {
    if (daySet.has(day)) {
      dp[day] = Math.min(dp[day + 1] + a, dp[day + 7] + b, dp[day + 30] + c)
    } else {
      dp[day] = dp[day + 1]
    }
  }

  return dp[0]
}
